Preliminaries for matching with memoization

mjambon opened the pull request:
This branch adds two main things:
* updated and renamed test suite for matching performance (`semgrep-core/perf`)
* toy sequence matcher

The toy matcher operates on sequences of chars, and matches a pattern over the entirety of the input sequence. See the readme for the motivations and usage.

I added optional memoization of the recursive match function, which gives satisfying results (30x speedup on the `quadratic` test, 600x speedup on the `cubic` test). The difficulty in the implementation is to ensure that the full environment is captured in the cache key and then to use suitable `hash` and `equal` functions. Creating cache keys and looking up cache entries is expensive, and in this implementation it turned out that only caching a random 1/3 of all computations was 1.5x faster than caching them all.

These tests run as part of semgrep-core's `make test`.

Plan for semgrep-core:

* port the kind of memoization we did here to semgrep-core statements.
* leave memoization optional to make performance comparisons easier - at least initially.

The plan for spacegrep is similar. It may be needed even more than for semgrep-core, but spacegrep currently avoids problems by limiting the span of ellipses to 10 lines (task #2255).

Commit:
Add toy matcher with a very first test

Commit:
Some tests for the toy matcher. Needs fixing.

Commit:
Add tracing of match calls (up to 100 calls)

Commit:
Fix ellipsis matching

Commit:
Add optional caching (memoization) to the toy matcher.
Preliminary results:
- the same test makes only 40,000 calls to the match function instead of
  50,000,000 without caching.
- the cached version takes slightly longer!

Commit:
Use more realistic (= non-uniform) test input, resulting in better
performance.

Commit:
Only consult the cache on 1/3 of the inputs, making things a bit faster.

Commit:
Add readme to toy matcher + add to semgrep-core 'make test'

Commit:
Skip slow benchmark tests by default

Commit:
Unskip a test, eliminate semgrep warning

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+     Some (Some ("foo", acc)) = acc is the stack of symbols, i.e. the list
+     of symbols captured so far by the current ellipsis in reverse order
+*)
+type ellipsis = (string * char list) option option
```
maybe better to define a new type for those 3 variants instead of using option option.
type capture_ellipsis = NotInEllipsis | EllipsisButNoCapture | EllipsisCapture of string * char list

mjambon replied:
I did that.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+open Printf
+
+type pattern_atom =
+  | Any_symbol         (* match any single symbol *)
```
Why not Any_char? Why symbol?

aryx replied:
Why not closer to what semgrep does and use:
type pattern =
  | Literal of char
  | Metavar of string
  | Ellipsis

instead of split in 2 with pattern_atom and pattern with those option string.
Also why do you need to capture the content of an Ellipsis?
Is it to imitate the ugly thing I do in semgrep-core to capture the range of the sequence of matched statements?

mjambon replied:
> Why not Any_char? Why symbol?

It's because we don't really care that they are of type `char`. It could be anything.

> Why not closer to what semgrep does and use:
> type pattern =
> | Literal of char
> | Metavar of string
> | Ellipsis
>
> instead of split in 2 with pattern_atom and pattern with those option string.

I picked the list because the list type is well-known. It makes it clear that the pattern is structured as a list rather than any tree.

> Also why do you need to capture the content of an Ellipsis?
> Is it to imitate the ugly thing I do in semgrep-core to capture the range of the sequence of matched statements?

Yes. Like anything that determines the result of the function, it must be made part of the cache key.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+  mutable match_calls: int;
+}
+
+let sample_pattern : pattern = [
```
Then this could be just:
let sampe_pattern = [
  Literal 'A'; Metavar "thing"; Ellipsis; Metavar "thing"]

mjambon replied:
I think it's clearer this way. I added comments about the differences between semgrep and this.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+
+(*
+   Unlike in semgrep syntax, we distinguish metavariable assignments
+   from metavariable dereferencing.
```
why?

aryx replied:
To make things more explicit? What do you win by having Backref and the string option in pattern to
represent those 2 different things?

mjambon replied:
I explained this in the file.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+   Initialize the capture of symbols by an ellipsis.
+   'init_acc' is a stack, holding symbols in reverse order.
+*)
+let init_ellipsis opt_name init_acc =
```
this would be clearer if we had 3 constructors instead of Some Some.

mjambon replied:
I did that.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+let extend env opt_ellipsis opt_name captured_sequence =
+  let env =
+    match opt_ellipsis with
+    | None -> env       (* we were not in an ellipsis *)
```
You would not need those extra comments if None was NotInEllipsis.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+      printf "[exceeded max trace lines = %i]\n%!" max_trace_lines
+
+module Cache_key = struct
+  type t = ellipsis * env * pattern * input
```
Can you add a comment explaining the intuition behind this caching idea. What do you try to cache
and why it avoids work.

mjambon replied:
I added  something.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+    trace_match_call ~trace stat ellipsis env pat input;
+    !get_from_cache ellipsis env pat input
+
+  and uncached_match ellipsis env pat input =
```
would be nice  to have the return type here. What does a matcher return?

aryx replied:
Apparently either None when no match or a Some with a triplet with X, the environment, and the rest of the input.

mjambon replied:
I clarified this

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+
+   Checks if a pattern matches the entire input sequence.
+   Returns the captured symbols or sequences of symbols for which a
+   name was specified in the pattern.
```
Maybe a small comment saying that essentially the code below
recurse on the pattern list, and inside this recurse on the input list.

mjambon replied:
done

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+                  | _ ->
+                      None
+                 )
+             | _ -> None
```
Maybe better to be explicit here, given there's not so many cases in pattern_atom:
| Any_symbol  | Symbol _ -> None

mjambon replied:
done

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+                  None
+  in
+  if cache then
+    get_from_cache := Memoize.create uncached_match
```
Would be great to have an example where I can understand how memoizing would improve the performance.
On your sample_pattern, with a simple input like A B C C C C C C C C B C C
why do we avoid redoing the same work thx to memoization?

mjambon replied:
This is a lot harder to explain than I thought. See other my comment. Best illustration is a trace.

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+  )
+
+(*
+   This is equivalent in semgrep to searching for a pattern like
```
This is great!

aryx commented on semgrep-core/toy-matcher/Matcher.ml:
```diff
@@ -0,0 +1,562 @@
+   This is equivalent in semgrep to searching for a pattern like
+   '$A; ... foo;', in a file where the statement 'foo;' doesn't exist.
+
+   With the naive match algorithm, the complexity is O(n^2) where n is the
```
can you explain why it's n^2, by unfolding the work done each time.

mjambon replied:
I added a bunch of comments to the file. For this specifically, I tried explaining things in comments but concluded it's better to run `make bench` and observe the traces.

aryx reviewed (changes requested):
This looks good. I would just like a few more comments.

Commit:
Clarify ellipsis accumulator type

Commit:
Add documentation and to-do

Commit:
Better style

Commit:
Documentation

mjambon commented:
I made various changes, mostly to clarify things.

There's just one thing that's not working, I left todo notes. Essentially, a semgrep pattern like `$A; ... foo($B);` can benefit from caching but not `$A; ... foo($A);` because in the second case, the value captured by `$A` is used in the tail of the pattern. Right now the toy matcher won't benefit even in the first case. This is because `$A` isn't discarded from the environment when constructing the cache keys for input matched against `... foo($B)`. (Caching does work in some of the the tests since we support matching any symbol without capturing its value, which doesn't trigger the name pollution problem)

aryx reviewed (approved):
Better. Not sure what's going on with the CI though.

aryx commented:
Fix the ci so we can merge it.

Commit:
Fix toy-matcher makefile

aryx commented:
Not sure why you have all those pfff, ocaml-tree-sitter, and ocaml-tree-sitter-lang modifications. Maybe merge develop first?

Commit:
Carry around minimal environment in the toy matcher so as to minimize
reuse of cache entries, whose keys include this set of bindings.

Commit:
Document toy matcher

Commit:
Prepare pattern for cached matching (needs a few unit tests)

Commit:
Add 'Caching.prepare_target'

aryx commented on semgrep-core/parsing/Caching.ml:
```diff
@@ -0,0 +1,163 @@
+    Visitor_AST.default_visitor with
+    kident = (fun (_k, _) (id, _tok) ->
+      if debug then
+        printf "kident %s\n" id;
```
You can use the Logging module in pfff/commons/. It's not too bad. Look for the use of logger in other files.

aryx reviewed (approved):
Ok let's merge this because this is getting really big.

aryx committed:
Merge branch 'develop' into mj/match-perf

aryx committed:
Disable calls to Caching.prepare_xxx because of many test regressions.
Probably comes from the need to provide an empty equal on s_backrefs
like we have for s_bf.

mjambon commented:
I forgot we had a pending pull request for this. Sorry about that. I kept adding changes to it and it wasn't in a mergeable state yesterday. It looks like you found a way around the errors, which is good.
